### PR TITLE
PERP-4486 | Add test for rujira zero price

### PR DESCRIPTION
**Pull request by Nomad-Free-Talent**
[PERP-4486](https://phobosfinance.atlassian.net/browse/PERP-4486)

[PERP-4486]: https://phobosfinance.atlassian.net/browse/PERP-4486?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by snoyberg — CHANGES_REQUESTED** [review]
We need a test showing what happens in the market contract itself. From the Jira issue:

> In other words, no positions should be opened or closed, liquifunding shouldn’t occur, etc. More directly: don’t wreck longs and reward shorts just because the feed says 0.

The point is to test the behavior of the entire market contract, not unit testing one piece. Please follow the rest of the integration test suite.

**Comment by Nomad-Free-Talent** [comment]
@snoyberg I spent several hours trying to add the multi-test for the zero-price issue from the rujira oracle.
It seems like the pre-defined util methods which is available in the current codebase can't be used for the zero-price issue because the methods like `exec_set_price` only receives `NonZero` parameter.
I think that the only available method is mocking the grpc response which I did for the unit test in this PR. However, I didn't find a proper solution to inject this test code *conditionally* to the existing multi-test code.
Can you guide me to add the custom grpc response logic conditionally to the multi-test package, please?

**Comment by snoyberg** [comment]
No, I can't, I don't know how to do this without diving in deeper myself, which I don't have time for now. I do expect that you'll  need to mock the response.

File: packages/multi_test/Cargo.toml [hunk 1]
```diff
@@ -30,6 +30,15 @@ cw2 = "2.0.0"
...
 dotenv = "0.15.0"
 cw-storage-plus = { version = "2.0.0", default-features = false }
+rujira-rs = { git = "https://gitlab.com/thorchain/rujira.git", rev = "acdbaa19836bb9cfa2e62ada6e8a2227373fe9df" }
+tokio = { version = "1.44.2", features = ["full"] }
+tokio-util = { version = "0.7", features = ["full"] }
+test-log = { version = "0.2", features = ["trace"] }
+pyth-sdk-cw = { workspace = true}
+
+[dev-dependencies]
+mocktail = "0.2.4-alpha"
```
Comment (snoyberg) [inline packages/multi_test/Cargo.toml:40]: Are we using this in the end? And can you go over the other added deps and confirm if they need to be present or not. I'd also like to know if we need to upgrade the Rust version or not, I'd rather avoid doing so if it's not needed.

File: packages/multi_test/tests/multi_test/countertrade.rs [hunk 2]
```diff
@@ -1793,6 +1793,7 @@ fn open_position_no_balance_mismatch_on_failure() {
     let work = market.query_countertrade_work().unwrap();
     assert!(work.is_open_position());
     market.exec_countertrade_do_work().unwrap_err();
+    std::env::remove_var("LEVANA_CONTRACTS_INJECT_FAILURE");
```
Comment (snoyberg) [inline packages/multi_test/tests/multi_test/countertrade.rs:1796]: This looks surprising, is it needed?

File: packages/multi_test/tests/multi_test/liquidation_price.rs [hunk 3]
```diff
@@ -292,3 +296,52 @@ fn normal_price_trader_gets_something() {
     assert!(closed.active_collateral > Collateral::zero());
     assert!(closed.pnl_collateral > "-100".parse().unwrap());
 }
+
+#[test]
+fn rujira_zero_price_logic_to_prevent_liquidation_perp_4486() {
```
Comment (snoyberg) [inline packages/multi_test/tests/multi_test/liquidation_price.rs:301]: What is this test case adding to our test coverage?

File: packages/multi_test/tests/multi_test/rujira_invalid_price/helper.rs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+use rujira_rs::proto::types::{QueryPoolRequest, QueryPoolResponse};
+
+pub const FACTORY_ADDR: Item<Addr> = Item::new(namespace::FACTORY_ADDR);
+
+pub const _GOVERNANCE: &str = "cosmwasm1h72z9g4qf2kjrq866zgn78xl32wn0q8aqayp05jkjpgdp2qft5aquanhrh";
+pub const MARKET_ADDR1: &str =
+    "cosmwasm1qnufjmd8vwm6j6d3q28wxqr4d8408f34fpka4vs365fvskualrasv5ues5";
+pub const _MARKET_ADDR2: &str =
+    "cosmwasm1vqjarrly327529599rcc4qhzvhwe34pp5uyy4gylvxe5zupeqx3sg08lap";
+pub struct CustomGrpcQuerier {
```
Comment (snoyberg) [inline packages/multi_test/tests/multi_test/rujira_invalid_price/helper.rs:18]: Can we have just one implementation of this instead of multiple copies in the code?

File: packages/multi_test/Cargo.toml (hunk 1 above)
Comment (Nomad-Free-Talent) [reply packages/multi_test/Cargo.toml:40]: As far as we decided not to use the mock grpc server anymore, I'm removing this one.

File: packages/multi_test/tests/multi_test/countertrade.rs (hunk 2 above)
Comment (Nomad-Free-Talent) [reply packages/multi_test/tests/multi_test/countertrade.rs:1796]: This is not related with our current task.
However, I found that the env variable is set from [line 1790](https://github.com/Levana-Protocol/levana-perps/pull/1080/files/8f0fe74087b4a2cfcd606e9b8536be5f3732bd55#diff-0d6935185da9d2ae7676dda584a6eafddd79fc4fd0480eb4b7b9458c4de0443bR1790) but not removed.
For the other places where this variable is set, they all got removed before end the test case.

**Comment by snoyberg** [comment]
Closing as wontfix, we ended up testing this extensively on a live chain.
